File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/AutoClosingRoomOpenHelperTest.java
/*
 * Copyright 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.room.integration.testapp.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import android.content.Context;
import android.database.Cursor;

import androidx.annotation.NonNull;
import androidx.arch.core.executor.testing.CountingTaskExecutorRule;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.testing.TestLifecycleOwner;
import androidx.room.InvalidationTracker;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.integration.testapp.TestDatabase;
import androidx.room.integration.testapp.dao.UserDao;
import androidx.room.integration.testapp.vo.User;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.filters.MediumTest;
import androidx.testutils.AssertionsKt;

import org.jetbrains.annotations.NotNull;
import org.junit.After;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;

import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;

// TODO: Consolidate with AutoClosingDatabaseTest that has access to internal APIs.
public class AutoClosingRoomOpenHelperTest {
    @Rule
    public CountingTaskExecutorRule mExecutorRule = new CountingTaskExecutorRule();
    private UserDao mUserDao;
    private TestDatabase mDb;
    private final DatabaseCallbackTest.TestDatabaseCallback mCallback =
            new DatabaseCallbackTest.TestDatabaseCallback();

    @Before
    public void createDb() throws TimeoutException, InterruptedException {
        Context context = ApplicationProvider.getApplicationContext();
        context.deleteDatabase("testDb");
        mDb = Room.databaseBuilder(context, TestDatabase.class, "testDb")
                .setAutoCloseTimeout(10, TimeUnit.MILLISECONDS)
                .addCallback(mCallback).build();
        mUserDao = mDb.getUserDao();
    }

    @After
    public void cleanUp() throws Exception {
        drain();
        mDb.close();
    }

    @Test
    @MediumTest
    public void inactiveConnection_shouldAutoClose() throws Exception {
        assertFalse(mCallback.mOpened);
        User user = TestUtil.createUser(1);
        user.setName("bob");
        mUserDao.insert(user);
        assertTrue(mCallback.mOpened);

        Thread.sleep(100);
        // Connection should be auto closed here

        User readUser = mUserDao.load(1);
        assertEquals(readUser.getName(), user.getName());
    }

    @Test
    @MediumTest
    public void slowTransaction_keepsDbAlive() throws Exception {
        assertFalse(mCallback.mOpened);

        User user = TestUtil.createUser(1);
        user.setName("bob");
        mUserDao.insert(user);
        assertTrue(mCallback.mOpened);
        Thread.sleep(30);
        mUserDao.load(1);
        // Connection should be auto closed here
        mDb.runInTransaction(
                () -> {
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException ignored) { }
                    // Connection would've been auto closed here
                }
        );

        Thread.sleep(100);
        // Connection should be auto closed here
    }

    @Test
    @MediumTest
    public void slowCursorClosing_keepsDbAlive() throws Exception {
        User user = TestUtil.createUser(1);
        user.setName("bob");
        mUserDao.insert(user);
        mUserDao.load(1);

        Cursor cursor = mDb.query("select * from user", null);

        Thread.sleep(100);

        cursor.close();

        Thread.sleep(100);
        // Connection should be auto closed here
    }

    @Test
    @MediumTest
    public void autoClosedConnection_canReopen() throws Exception {
        User user1 = TestUtil.createUser(1);
        user1.setName("bob");
        mUserDao.insert(user1);

        Thread.sleep(100);
        // Connection should be auto closed here

        User user2 = TestUtil.createUser(2);
        user2.setName("bob2");
        mUserDao.insert(user2);
        Thread.sleep(100);
        // Connection should be auto closed here
    }

    @Test
    @MediumTest
    public void liveDataTriggers_shouldApplyOnReopen() throws Exception {
        LiveData<Boolean> adminLiveData = mUserDao.isAdminLiveData(1);

        final TestLifecycleOwner lifecycleOwner = new TestLifecycleOwner();
        final TestObserver<Boolean> observer = new AutoClosingRoomOpenHelperTest
                .MyTestObserver<>();
        TestUtil.observeOnMainThread(adminLiveData, lifecycleOwner, observer);
        assertNull(observer.get());

        User user = TestUtil.createUser(1);
        user.setAdmin(true);
        mUserDao.insert(user);

        assertNotNull(observer.get());
        assertTrue(observer.get());

        user.setAdmin(false);
        mUserDao.insertOrReplace(user);
        assertNotNull(observer.get());
        assertFalse(observer.get());

        Thread.sleep(100);
        // Connection should be auto closed here

        user.setAdmin(true);
        mUserDao.insertOrReplace(user);
        assertNotNull(observer.get());
        assertTrue(observer.get());
    }

    @Test
    @MediumTest
    public void testCanExecSqlInCallback() throws Exception {
        Context context = ApplicationProvider.getApplicationContext();
        context.deleteDatabase("testDb2");
        TestDatabase db = Room.databaseBuilder(context, TestDatabase.class, "testDb2")
                        .setAutoCloseTimeout(10, TimeUnit.MILLISECONDS)
                        .addCallback(new ExecSqlInCallback())
                        .build();

        db.getUserDao().insert(TestUtil.createUser(1));

        db.close();
    }

    @Test
    public void testManuallyRoomDatabaseClose() throws Exception {
        Context context = ApplicationProvider.getApplicationContext();
        // Create a new db since the other one is cleared in the @After
        TestDatabase testDatabase = Room.databaseBuilder(context, TestDatabase.class, "testDb2")
                .setAutoCloseTimeout(10, TimeUnit.MILLISECONDS)
                .addCallback(new ExecSqlInCallback())
                .build();

        testDatabase.close();

        // We shouldn't be able to do anything with the database now...
        AssertionsKt.assertThrows(IllegalStateException.class, () -> {
            testDatabase.getUserDao().count();
        }).hasMessageThat().contains("closed");

        assertFalse(testDatabase.isOpen());

        assertFalse(testDatabase.isOpen());
        TestDatabase testDatabase2 = Room.databaseBuilder(context, TestDatabase.class, "testDb2")
                .setAutoCloseTimeout(10, TimeUnit.MILLISECONDS)
                .addCallback(new ExecSqlInCallback())
                .build();
        testDatabase2.getUserDao().count(); // db should open now
        testDatabase2.close();
        assertFalse(testDatabase.isOpen());
    }

    @Test
    public void testManuallyOpenHelperClose() throws Exception {
        Context context = ApplicationProvider.getApplicationContext();
        // Create a new db since the other one is cleared in the @After
        TestDatabase testDatabase = Room.databaseBuilder(context, TestDatabase.class, "testDb2")
                .setAutoCloseTimeout(10, TimeUnit.MILLISECONDS)
                .addCallback(new ExecSqlInCallback())
                .build();

        testDatabase.getOpenHelper().close();

        // We shouldn't be able to do anything with the database now...
        AssertionsKt.assertThrows(IllegalStateException.class, () -> {
            testDatabase.getUserDao().count();
        }).hasMessageThat().contains("closed");

        assertFalse(testDatabase.isOpen());
        TestDatabase testDatabase2 = Room.databaseBuilder(context, TestDatabase.class, "testDb2")
                .setAutoCloseTimeout(10, TimeUnit.MILLISECONDS)
                .addCallback(new ExecSqlInCallback())
                .build();
        testDatabase2.getUserDao().count(); // db should open now
        testDatabase2.getOpenHelper().close();
        assertFalse(testDatabase.isOpen());
    }

    // TODO(336671494): broken test
    @Ignore
    @Test
    @MediumTest
    public void invalidationObserver_isCalledOnEachInvalidation()
            throws TimeoutException, InterruptedException {
        AtomicInteger invalidationCount = new AtomicInteger(0);

        UserTableObserver userTableObserver =
                new UserTableObserver(invalidationCount::getAndIncrement);

        mDb.getInvalidationTracker().addObserver(userTableObserver);

        mUserDao.insert(TestUtil.createUser(1));

        drain();
        assertEquals(1, invalidationCount.get());

        User user1 = TestUtil.createUser(1);
        user1.setAge(123);
        mUserDao.insertOrReplace(user1);

        drain();
        assertEquals(2, invalidationCount.get());

        Thread.sleep(15);
        // Connection should be closed now

        mUserDao.insert(TestUtil.createUser(2));

        drain();
        assertEquals(3, invalidationCount.get());
    }

    // TODO(372946311): Broken test
    @Ignore
    @Test
    @MediumTest
    public void invalidationObserver_canRequeryDb() throws TimeoutException, InterruptedException {
        Context context = ApplicationProvider.getApplicationContext();

        context.deleteDatabase("testDb2");
        TestDatabase db = Room.databaseBuilder(context, TestDatabase.class, "testDb2")
                // create contention for callback
                .setAutoCloseTimeout(0, TimeUnit.MILLISECONDS)
                .addCallback(mCallback).build();

        AtomicInteger userCount = new AtomicInteger(0);

        UserTableObserver userTableObserver = new UserTableObserver(
                () -> userCount.set(db.getUserDao().count()));

        db.getInvalidationTracker().addObserver(userTableObserver);

        db.getUserDao().insert(TestUtil.createUser(1));
        db.getUserDao().insert(TestUtil.createUser(2));
        db.getUserDao().insert(TestUtil.createUser(3));
        db.getUserDao().insert(TestUtil.createUser(4));
        db.getUserDao().insert(TestUtil.createUser(5));
        db.getUserDao().insert(TestUtil.createUser(6));
        db.getUserDao().insert(TestUtil.createUser(7));

        drain();
        assertEquals(7, userCount.get());
        db.close();
    }

    private void drain() throws TimeoutException, InterruptedException {
        mExecutorRule.drainTasks(1, TimeUnit.MINUTES);
    }

    private class MyTestObserver<T> extends TestObserver<T> {
        @Override
        protected void drain() throws TimeoutException, InterruptedException {
            AutoClosingRoomOpenHelperTest.this.drain();
        }
    }

    private static class ExecSqlInCallback extends RoomDatabase.Callback {
        @Override
        public void onOpen(@NonNull SupportSQLiteDatabase db) {
            db.query("select * from user").close();
        }
    }

    private static class UserTableObserver extends InvalidationTracker.Observer {

        private final Runnable mInvalidationCallback;

        UserTableObserver(Runnable invalidationCallback) {
            super("user");
            mInvalidationCallback = invalidationCallback;
        }

        @Override
        public void onInvalidated(@NonNull @NotNull Set<String> tables) {
            mInvalidationCallback.run();
        }
    }
}

File: room/room-runtime/src/androidTest/java/androidx/room/AutoClosingRoomOpenHelperTest.kt
/*
 * Copyright 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.room

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteException
import android.os.Build
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.sqlite.db.SupportSQLiteOpenHelper
import androidx.sqlite.db.framework.FrameworkSQLiteOpenHelperFactory
import androidx.test.core.app.ApplicationProvider
import androidx.test.filters.FlakyTest
import androidx.test.filters.SdkSuppress
import androidx.testutils.assertThrows
import com.google.common.truth.Truth.assertThat
import org.junit.Before
import org.junit.Test
import java.io.IOException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

public class AutoClosingRoomOpenHelperTest {

    private open class Callback(var throwOnOpen: Boolean = false) :
        SupportSQLiteOpenHelper.Callback(1) {
        override fun onCreate(db: SupportSQLiteDatabase) {}

        override fun onOpen(db: SupportSQLiteDatabase) {
            if (throwOnOpen) {
                throw IOException()
            }
        }

        override fun onUpgrade(db: SupportSQLiteDatabase, oldVersion: Int, newVersion: Int) {}
    }

    @Before
    public fun setUp() {
        ApplicationProvider.getApplicationContext<Context>().deleteDatabase("name")
    }

    private fun getAutoClosingRoomOpenHelper(
        timeoutMillis: Long = 10,
        callback: SupportSQLiteOpenHelper.Callback = Callback()
    ): AutoClosingRoomOpenHelper {

        val delegateOpenHelper = FrameworkSQLiteOpenHelperFactory()
            .create(
                SupportSQLiteOpenHelper.Configuration
                    .builder(ApplicationProvider.getApplicationContext())
                    .callback(callback)
                    .name("name")
                    .build()
            )

        val autoCloseExecutor = Executors.newSingleThreadExecutor()

        return AutoClosingRoomOpenHelper(
            delegateOpenHelper,
            AutoCloser(timeoutMillis, TimeUnit.MILLISECONDS, autoCloseExecutor).apply {
                init(delegateOpenHelper)
                setAutoCloseCallback { }
            }
        )
    }

    @Test
    public fun testQueryFailureDecrementsRefCount() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()

        assertThrows<SQLiteException> {
            autoClosingRoomOpenHelper
                .writableDatabase.query("select * from nonexistanttable")
        }

        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(0)
    }

    @Test
    public fun testCursorKeepsDbAlive() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()
        autoClosingRoomOpenHelper.writableDatabase.execSQL("create table user (idk int)")

        val cursor =
            autoClosingRoomOpenHelper.writableDatabase.query("select * from user")
        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(1)
        cursor.close()
        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(0)
    }

    @Test
    public fun testTransactionKeepsDbAlive() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()
        autoClosingRoomOpenHelper.writableDatabase.beginTransaction()
        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(1)
        autoClosingRoomOpenHelper.writableDatabase.endTransaction()
        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(0)
    }

    @Test
    @SdkSuppress(minSdkVersion = Build.VERSION_CODES.JELLY_BEAN)
    public fun enableWriteAheadLogging_onOpenHelper() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()

        autoClosingRoomOpenHelper.setWriteAheadLoggingEnabled(true)
        assertThat(autoClosingRoomOpenHelper.writableDatabase.isWriteAheadLoggingEnabled).isTrue()

        Thread.sleep(100) // Let the db auto close...

        assertThat(autoClosingRoomOpenHelper.writableDatabase.isWriteAheadLoggingEnabled).isTrue()
    }

    @Test
    public fun testEnableWriteAheadLogging_onSupportSqliteDatabase_throwsUnsupportedOperation() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()

        assertThrows<UnsupportedOperationException> {
            autoClosingRoomOpenHelper.writableDatabase.enableWriteAheadLogging()
        }

        assertThrows<UnsupportedOperationException> {
            autoClosingRoomOpenHelper.writableDatabase.disableWriteAheadLogging()
        }
    }

    @FlakyTest(bugId = 190607416)
    @Test
    public fun testOnOpenCalledOnEachOpen() {
        val countingCallback = object : Callback() {
            var onCreateCalls = 0
            var onOpenCalls = 0

            override fun onCreate(db: SupportSQLiteDatabase) {
                onCreateCalls++
            }

            override fun onOpen(db: SupportSQLiteDatabase) {
                super.onOpen(db)
                onOpenCalls++
            }
        }

        val autoClosingRoomOpenHelper =
            getAutoClosingRoomOpenHelper(callback = countingCallback)

        autoClosingRoomOpenHelper.writableDatabase
        assertThat(countingCallback.onOpenCalls).isEqualTo(1)
        assertThat(countingCallback.onCreateCalls).isEqualTo(1)

        Thread.sleep(20) // Database should auto-close here
        autoClosingRoomOpenHelper.writableDatabase
        assertThat(countingCallback.onOpenCalls).isEqualTo(2)
        assertThat(countingCallback.onCreateCalls).isEqualTo(1)
    }

    @Test
    public fun testStatementReturnedByCompileStatement_doesntKeepDatabaseOpen() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()

        val db = autoClosingRoomOpenHelper.writableDatabase
        db.execSQL("create table user (idk int)")

        db.compileStatement("insert into users (idk) values (1)")

        Thread.sleep(20)
        assertThat(db.isOpen).isFalse() // db should close
        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(0)
    }

    @Test
    public fun testStatementReturnedByCompileStatement_reOpensDatabase() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()

        val db = autoClosingRoomOpenHelper.writableDatabase
        db.execSQL("create table user (idk int)")

        val statement = db
            .compileStatement("insert into user (idk) values (1)")

        Thread.sleep(20)

        statement.executeInsert() // This should succeed

        db.query("select * from user").useCursor {
            assertThat(it.count).isEqualTo(1)
        }

        assertThat(autoClosingRoomOpenHelper.autoCloser.refCountForTest).isEqualTo(0)
    }

    @Test
    public fun testStatementReturnedByCompileStatement_worksWithBinds() {
        val autoClosingRoomOpenHelper = getAutoClosingRoomOpenHelper()
        val db = autoClosingRoomOpenHelper.writableDatabase

        db.execSQL("create table users (i int, d double, b blob, n int, s string)")

        val statement = db.compileStatement(
            "insert into users (i, d, b, n, s) values (?,?,?,?,?)"
        )

        statement.bindString(5, "123")
        statement.bindLong(1, 123)
        statement.bindDouble(2, 1.23)
        statement.bindBlob(3, byteArrayOf(1, 2, 3))
        statement.bindNull(4)

        statement.executeInsert()

        db.query("select * from users").useCursor {
            assertThat(it.moveToFirst()).isTrue()
            assertThat(it.getInt(0)).isEqualTo(123)
            assertThat(it.getDouble(1)).isWithin(.01).of(1.23)

            assertThat(it.getBlob(2)).isEqualTo(byteArrayOf(1, 2, 3))
            assertThat(it.isNull(3)).isTrue()
            assertThat(it.getString(4)).isEqualTo("123")
        }

        statement.clearBindings()
        statement.executeInsert() // should insert with nulls

        db.query("select * from users").useCursor {
            assertThat(it.moveToFirst()).isTrue()
            it.moveToNext()
            assertThat(it.isNull(0)).isTrue()
            assertThat(it.isNull(1)).isTrue()
            assertThat(it.isNull(2)).isTrue()
            assertThat(it.isNull(3)).isTrue()
            assertThat(it.isNull(4)).isTrue()
        }
    }

    @Test
    public fun testGetDelegate() {
        val delegateOpenHelper = FrameworkSQLiteOpenHelperFactory()
            .create(
                SupportSQLiteOpenHelper.Configuration
                    .builder(ApplicationProvider.getApplicationContext())
                    .callback(Callback())
                    .name("name")
                    .build()
            )

        val autoCloseExecutor = Executors.newSingleThreadExecutor()

        val autoClosing = AutoClosingRoomOpenHelper(
            delegateOpenHelper,
            AutoCloser(0, TimeUnit.MILLISECONDS, autoCloseExecutor)
        )

        assertThat(autoClosing.getDelegate()).isSameInstanceAs(delegateOpenHelper)
    }

    // Older API versions didn't have Cursor implement Closeable
    private inline fun Cursor.useCursor(block: (Cursor) -> Unit) {
        try {
            block(this)
        } finally {
            this.close()
        }
    }
}